### ISSUE TITLE
[FEATURE_REQUEST] Story Writing/Text Editor Mode

### ISSUE THREAD

**Issue by engadine1997** [SillyTavern/SillyTavern#2516]
### Have you searched for similar requests?

Yes

### Describe the solution you'd like

As stated in the title, though I understand if this is not within the scope of this project.

SillyTavern is in my opinion the best chat-oriented frontend in terms of features and customisability, but I think it would benefit from a story writing mode that would change the chat interface into a text editor. Something that lets the LLM act as a writing assistant, continuing a document of text directly with the occasional guidance as user instruct messages, with features like a lorebook working in the background. Similar examples can be seen in a frontend like Mikupad or a paid service like NovelAI.

### Describe alternatives you've considered

- Can be crudely achieved with a blank character card, story writing system prompt, lorebook for information, and the 'Document' chat style. However, this just gives the illusion of a continuous document as subsequent generations just send a whole new AI instruct message (which can confuse the AI by splitting things over messages). Editing/continuing a single AI generation to make it write continuously is also quite cumbersome.

- Alternative projects exist, but they are quite lacking in every other regard (customisability, general user experience, utilities like vector storage or lorebook)

### Priority

Low (Nice-to-have)

### Are you willing to test this on staging/unstable branch if this is implemented?

Yes

**Comment by Cohee1207** [comment]
It's not impossible, but requires substantial frontend changes and a completely new prompt builder or hacking into the existing one as right now messages that get into the context are not split. That would require treating every paragraph as a virtual "message" for the prompt builder to consider. Not to mention this begs to completely rethink swipes as an entity.

Also, keep in mind one fact that everyone basically overlooks - having a "story mode" basically means no markdown formatting for the text unless there's a comprehensive enough WYSIWYG editor that can convert to and from markdown seamlessly - the existing chat paradigm with user input and display being separate will not work.

**Comment by engadine1997** [comment]
You're right about the whole architectural issues, and I hadn't really considered the problems with adapting the current chat paradigm. The implementation I was picturing was something akin to [mikupad](https://github.com/lmg-anon/mikupad) - essentially expose the underlying context for editing.

Mikupad does not keep track of messages or swipes, and instead lets the user edit the entire raw context window that gets send to the model. Generation appends to the document when the model streams a prediction (though it must be doing something extra to let you undo or redo a generation). It's not truly a "story mode" with separate user input and display as you mentioned - in hindsight I realise that would be difficult to implement - but this approach makes it convenient to quickly rewrite parts of AI generations.

To get an instruct model to follow an instruction in Mikupad's "raw" approach, you have to type the literal instruct sequence into the document so it can continue the 'AI turn' message. This isn't particularly elegant as writing the instruct sequence manually is both tedious and ugly, but perhaps there are better ways to go about this. Maybe the instruct tokens/template could be hidden from the user, and the current send message field could be adapted to append an instruction to the document and automatically surround it with the appropriate instruct scaffold.

> "[...] having a "story mode" basically means no markdown formatting for the text [...]"

As for the lack of markdown you mentioned, yes a completely raw approach like this wouldn't be able to display markdown. Is it even be possible to have a text editor render markdown continuously as it's being edited? In any case, I personally think not having markdown support wouldn't be a huge deal breaker for most (at least it's not for me). The only major use I can see it having in story writing is formatting text with strikethroughs, italics, etc. Image display too maybe, if that's how the img generation extension shows images in the chat currently.

Anyway, sorry for the long write-up 😅

**Comment by Wolfsblvt** [comment]
You could split the UI up differently, and have markdown being rendered on the right side next to what you are writing. Many markdown editors do that, and don't use WYSIWYG.
Not like here on GitHub, where it's just separate tabs, but displaying it next to each other.

I think that's the least concern for something like this.

**Comment by HonZuna** [comment]
Great idea.

**Comment by Wolfsblvt** [comment]
![image](https://github.com/user-attachments/assets/06a38d8a-cacc-46d3-82a9-bc0f57ed10fe)
VSCode with a markdown preview

**Comment by Lex-DRL** [comment]
+1 for a really desired feature, and here are my 5 cents to sweeten the deal. In essence, I believe ALMOST all functionality is already there. No need to change the architecture, only UI needs to be changed:
- as previously commented, you can just disable markdown formatting for story mode at the moment. Having story mode with raw text and no markdown is better than having no story mode at all.
- Keep the internal prompt generation as it is, but:
  - Global system prompt can be slightly tweaked to tell the model that chat below is a single story and the AI "should never, **UNDER NO CIRCUMSTANCES** address the user directly", treating their messages either as their part in continuing the story or a direct instruction on where the story should go next. I've tested this approach yesterday, and it works just fine with Dolphin-LLAMA-3.1.
  - QoL: Probably, there needs to be a hint for the user to create a special "writer" character with no lorebook attached, since this needs to be done upfront *(at least, I, as a newbie in ST, haven't found a way to "reassign" chat from one character to another)*.
    - Ideally, the writer char needs to be pre-made by default and there needs to be an option in settings which char is associated with story mode to auto-switch to chat with it whenever story mode is enabled.
  - Display the "chat" as a single wall of text. Visual separation lines between messages can be kept and probably are desired, as well as message buttons shown on hover. But there shouldn't be any avatars, big padding, etc. Though, it's desireable to have slightly different text color for AI/user messages. Basically, it's one of the styles already present in SillyTavern, except...
  - The entirety of the chat needs to be in a permanent "edit" mode, without the need to individually select > edit > confirm each message separately. And, obviously, cursor needs to go across the entire "chat log", freely jumping between messages as if they were a single one (but keeping under the hood which part of text belongs to which message).
  - When a text block (a "message") gets edited to become an empty string, the block itself needs to be auto-deleted, to prevent chat log from polluting with a ton of empty messages which (with their zero height) are hard to find.
  - A simple hotkey (Tab?) to switch between who a currently active block of text is written as. It could be useful when you want to steer the story in a certain direction by starting a sentence, and the AI needs to continue it as if it was it's own.
    - Optionally, all newly created text blocks should be created as the assistant's ones.
  - Ideally, regular Enter should create a new block ("message"), which might contain multiple paragraphs within it. And Shift+Enter should be treated as adding line breaks/paragraphs within an active block.
  - Ctr+Enter and Alt+Enter do the same thing as before. Though,
    - Ctrl+Enter should probably behave differently in story mode: it should create a new block and generate into it (i.e., the same thing a simple Enter does now).
    - It would be nice if there also was some hotkey (Ctrl+Alt+Enter?) to re-generate only the last generated part within current block (whatever was appended with last Alt+Enter). But I realize this would require changing the data model (text in messages would be an array of strings instead of a single string). As a workaround without changing the data model, message entities might contain an extra int field now, representing the number of characters (from the end of message) which were appended the last time. And when it's zero, it means the next Ctrl+Alt+Enter would re-generate the entire text block (though, to be safe, it should probably re-generate as swipe in such case).
- Optionally, you MIGHT add special state indicators for user messages and their special formatting within prompt. Like, auto-appending a prefix/postfix in prompt, which are different depending on whether its a story continuation "comment" or an instruction.
  - But, honestly, this is just a "nice to have", which can be safely delayed for future iterations on the mode. There's a ton of things that can be done automatically, like auto-marking user's instruction comments as hidden after a certain delay, but even without them, even a barebones story mode would be a HUUUGE improvement for those who use LLMs for writing and not for chatting.

And the best part, this solution is completely backwards-compatible with the regular chat mode, since it's only a different display of the same chat with the same prompt-building logic.

**Comment by Lex-DRL** [comment]
Also, I don't think a separate panel with markdown fits here. It's good for an IDE, but it takes A TON of precious screen real estate and might significantly slow the browser if it needs to update on each change of raw text.
It either should be done "the right way" (in-place, switchable back and forth with a raw mode, and very performant from the start) or it shouldn't be done at all.

... Or, at least, there needs to be an option to turn it off. Completely. With no perfomance hit compared to a case where this panel wasn't even implemented.
